include/realtime_tools/realtime_box_best_effort.h

```diff
@@ -0,0 +1,232 @@
+//
+//    * Neither the name of the Willow Garage, Inc. nor the names of its
+//      contributors may be used to endorse or promote products derived from
+//      this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+// Author: Lennart Nachtigall
+
+#ifndef REALTIME_TOOLS__REALTIME_BOX_BEST_EFFORT_H_
+#define REALTIME_TOOLS__REALTIME_BOX_BEST_EFFORT_H_
+
+#include <initializer_list>
+#include <mutex>
+#include <optional>
+#include <utility>
+
+#include <rcpputils/pointer_traits.hpp>
+
+namespace realtime_tools
+{
+
+template <typename T>
+constexpr auto is_ptr_or_smart_ptr = rcpputils::is_pointer<T>::value;
+
+/*!
+    A Box that ensures thread safe access to the boxed contents.
+    Access is best effort. If it can not lock it will return.
+
+    NOTE about pointers:
+    You can use pointers with this box but the access will be different.
+    Only use the get/set methods that take function pointer for accessing the internal value.
+*/
+template <class T, typename mutex_type = std::mutex>
+class RealtimeBoxBestEffort
+{
+  static_assert(
+    std::is_same_v<mutex_type, std::mutex> || std::is_same_v<mutex_type, std::recursive_mutex>);
+  static_assert(std::is_copy_constructible_v<T>, "Passed type must be copy constructible");
+
+public:
+  using mutex_t = mutex_type;
+  using type = T;
+  // Provide various constructors
+  constexpr explicit RealtimeBoxBestEffort(const T & init = T{}) : value_(init) {}
+  constexpr explicit RealtimeBoxBestEffort(const T && init) : value_(std::move(init)) {}
+
+  // Only enabled for types that can be constructed from an initializer list
+  template <typename U = T>
+  constexpr RealtimeBoxBestEffort(
+    const std::initializer_list<U> & init,
+    std::enable_if_t<std::is_constructible_v<U, std::initializer_list>>)
+  : value_(init)
+  {
+  }
+
+  /**
+   * @brief set a new content with best effort
+   * @return false if mutex could not be locked
+   * @note disabled for pointer types
+   */
+  template <typename U = T>
+  typename std::enable_if_t<!is_ptr_or_smart_ptr<U>, bool> trySet(const T & value)
+  {
+    std::unique_lock<std::mutex> guard(lock_, std::defer_lock);
```

Comment:
```suggestion
    std::unique_lock<mutex_t> guard(lock_, std::defer_lock);
```

Reply:
You are absolutely right about this one. Thats something I missed.